### ISSUE TITLE
Bad PAM configuration after installing from Ubuntu repository

### ISSUE THREAD

**Issue by bovender** [google/fscrypt#100]
I have installed fscrypt on Ubuntu 18.04:

    sudo apt install fscrypt # (no PPA required)
    sudo fscrypt setup

When I attempt to encrypt a directory, I get this:

    fscrypt encrypt: user keyring not linked into session keyring
    
    This is usually the result of a bad PAM configuration. Either correct the problem in your PAM stack, enable
    pam_keyinit.so, or run "keyctl link @u @s".

Contrary to what the README.md [suggests](https://github.com/google/fscrypt#automatic-setup-on-ubuntu), there is no `fscrypt` file in `/usr/share/pam-configs`, so I guess that's the problem.

Before I start messing about with my PAM configuration (something that I feel uncomfortable to do), I'd appreciate a pointer as to how to fix this, or maybe it could be fixed in the Ubuntu package?

Thanks.

**Comment by adembo** [comment]
@bovender I also ran into this. It turned out to be a problem in Ubuntu 18.04's PAM configuration for systemd that was already fixed upstream. See [the bug report](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1754270.) for details. There's a workaround posted there too.

**Comment by bovender** [comment]
Hm, the workaround does not work for me. Only when I issue `keyctl link @u @s` I can get it to work.

**Comment by kaxing** [comment]
Might be related, after `fscrypt setup` and `keyctl link @u @s` I couldn't opne X then found these in log:
Versions: Ubuntu 18.04, fscrypt 0.2.2-0ubuntu2

```
 PAM unable to dlopen(pam_kwallet.so): /lib/security/pam_kwallet.so: cannot open shared object file: No such file or directory
 PAM adding faulty module: pam_kwallet.so
 PAM unable to dlopen(pam_kwallet5.so): /lib/security/pam_kwallet5.so: cannot open shared object file: No such file or directory
 PAM adding faulty module: pam_kwallet5.so
 pam_unix(lightdm-greeter:session): session opened for user lightdm by (uid=0)
```

**Comment by PicoJr** [comment]
> @bovender I also ran into this. It turned out to be a problem in Ubuntu 18.04's PAM configuration for systemd that was already fixed upstream. See [the bug report](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1754270.) for details. There's a workaround posted there too.

Link is dead :/ I'm interested in the workaround ^^'

**Comment by tlbdk** [comment]
There is a dot at the end for the link: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1754270

**Comment by ebiggers** [comment]
This problem is still present on a newly installed Ubuntu 19.04 system after `sudo apt-get install fscrypt libpam-fscrypt`.

However, it's *not* present on a newly installed Ubuntu 20.04 (daily build) system.

It's not obvious what caused it to be fixed, though, as `pam_keyinit.so` wasn't added to any new PAM configuration files; it's still in the same ones as before.

**Comment by ebiggers** [comment]
We could also make `fscrypt` link the user's keyring into the current session keyring itself.  However it wouldn't solve the problem in all cases as we also support root unlocking a directory for another user.

Anyway, we'll also be switching to the new filesystem-level keyrings with Linux v5.4+ and #148, which avoids this problem.

**Comment by outusuke** [comment]
I solved my permission denied issue by adding

> Name: keyinit fix
Default: yes
Priority: 0
Session-Type: Additional
Session:
	optional	pam_keyinit.so force revoke

to /usr/share/pam-configs/keyinit-fix 
sudo apt pam-auth-update

I also noticed that pam_fscrypt goes to /usr/local/share/pam-configs folder but only works by moving to /usr/share/pam-configs

**Comment by ebiggers** [comment]
I just re-tested this on the latest versions of Ubuntu (20.04.2 and 21.04), and I wasn't able to reproduce this.  I tested both `custom_passphrase` and `pam_passphrase` protectors, and both worked fine.  (Note: for auto-unlocking to work with `pam_passphrase`,  it was necessary to install the `libpam-fscrypt` package, not just  `fscrypt`.)  So, I'm closing this issue, as it appears to have been fixed.
